client/src/lib/Router/Route.tsx

```diff
@@ -1,17 +1,68 @@
-import React from 'react';
-import { useRouterContext } from './Router';
+import React, { ReactNode } from 'react';
+import { IRouterContext, useRouterContext } from './Router';
 
 export interface IRouteProps {
   path: string;
   component: React.FC;
 }
 
-const Route = (props: IRouteProps): JSX.Element => {
+const compilePath = (path: string) => {
+  const keys: string[] = [];
+  const compiledPath = path.replace(/:(\w+)/g, (_, key) => {
+    keys.push(key);
+    return '([^\\/]+)';
+  });
+  const regex = new RegExp(`^(${compiledPath})$`, 'i');
+  return { regex, keys };
+};
+
+const checkRoute = (originPath: string, route: ReactNode, component: React.FC, routerCtx: IRouterContext) => {
+  const { path: targetPath } = (route as JSX.Element).props as IRouteProps;
+  const { currentPath } = routerCtx;
+
+  if (originPath === targetPath) {
+    if (targetPath === currentPath) {
+      return React.createElement(component);
+    }
+
+    const { regex, keys } = compilePath(targetPath);
+    const match = currentPath.match(regex);
+
+    if (match) {
+      const params = match.slice(2);
+      const routeParams = keys.reduce<{ [key: string]: string }>((obj, key, i) => {
+        obj[key] = params[i];
+        return obj;
+      }, {});
+
+      routerCtx.params = routeParams;
+      return React.createElement(component);
+    }
+  }
+
+  return null;
+};
+
+const Route = (props: IRouteProps) => {
   const { path, component } = props;
-  const { currentPath } = useRouterContext();
-  const isSamePath = path === currentPath;
+  const routerCtx = useRouterContext();
+  const { routes } = routerCtx;
+
+  if (Array.isArray(routes)) {
+    for (const route of routes) {
+      const result = checkRoute(path, route, component, routerCtx);
+      if (result) {
+        return result;
+      }
+    }
+  } else {
+    const result = checkRoute(path, routes, component, routerCtx);
+    if (result) {
+      return result;
+    }
+  }
```

Comment:
```suggestion 
  const router = useRouterContext();
  const { currentPath } = router;

  if (path === currentPath) {
    return React.createElement(component);
  }

  const { regex, keys } = compilePath(path);
  const match = currentPath.match(regex);
  if (match) {
    const params = match.slice(2);
    router.params = keys.reduce<{ [key: string]: string }>((obj, key, i) => {
      obj[key] = params[i];
      return obj;
    }, {});
    return React.createElement(component);
  }
```

전체 route를 순회하지 않아도 각 라우터가 리랜더링되며 내부 함수를 실행합니다.
순회가 사라진다면 이런 식으로도 표현할 수 있을 것 같아요 😄 
`router context`에 `routes`를 넘겨주지 않아도 될 것 같네요.